README.md
## Ethernet_Generic

[![arduino-library-badge](https://www.ardu-badge.com/badge/Ethernet_Generic.svg?)](https://www.ardu-badge.com/Ethernet_Generic)
[![GitHub release](https://img.shields.io/github/release/khoih-prog/Ethernet_Generic.svg)](https://github.com/khoih-prog/Ethernet_Generic/releases)
[![contributions welcome](https://img.shields.io/badge/contributions-welcome-brightgreen.svg?style=flat)](#Contributing)
[![GitHub issues](https://img.shields.io/github/issues/khoih-prog/Ethernet_Generic.svg)](http://github.com/khoih-prog/Ethernet_Generic/issues)

<a href="https://www.buymeacoffee.com/khoihprog6" title="Donate to my libraries using BuyMeACoffee"><img src="https://cdn.buymeacoffee.com/buttons/v2/default-yellow.png" alt="Donate to my libraries using BuyMeACoffee" style="height: 50px !important;width: 181px !important;" ></a>
<a href="https://www.buymeacoffee.com/khoihprog6" title="Donate to my libraries using BuyMeACoffee"><img src="https://img.shields.io/badge/buy%20me%20a%20coffee-donate-orange.svg?logo=buy-me-a-coffee&logoColor=FFDD00" style="height: 20px !important;width: 200px !important;" ></a>


---
---

## Table of Contents

* [Important Note from v2.5.2](#Important-Note-from-v252)
* [Why do we need this Ethernet_Generic library](#why-do-we-need-this-Ethernet_Generic-library)
  * [Features](#features)
  * [Currently supported Boards](#currently-supported-boards)
  * [Currently supported Ethernet shields/modules](#currently-supported-ethernet-shieldsmodules)
* [Changelog](changelog.md) 
* [Prerequisites](#prerequisites)
* [Installation](#installation)
  * [Use Arduino Library Manager](#use-arduino-library-manager)
  * [Manual Install](#manual-install)
  * [VS Code & PlatformIO](#vs-code--platformio)
* [Packages' Patches](#packages-patches)
  * [1. For Adafruit nRF52840 and nRF52832 boards](#1-for-adafruit-nRF52840-and-nRF52832-boards)
  * [2. For Teensy boards](#2-for-teensy-boards)
  * [3. For Arduino SAM DUE boards](#3-for-arduino-sam-due-boards)
  * [4. For Arduino SAMD boards](#4-for-arduino-samd-boards)
      * [For core version v1.8.10+](#for-core-version-v1810)
      * [For core version v1.8.9-](#for-core-version-v189-)
  * [5. For Adafruit SAMD boards](#5-for-adafruit-samd-boards)
  * [6. For Seeeduino SAMD boards](#6-for-seeeduino-samd-boards)
  * [7. For STM32 boards](#7-for-stm32-boards) 
    * [7.1. For STM32 boards to use LAN8720](#71-for-stm32-boards-to-use-lan8720)
    * [7.2. For STM32 boards to use Serial1](#72-for-stm32-boards-to-use-serial1)
  * [8. For RP2040-based boards using Earle Philhower arduino-pico core](#8-for-rp2040-based-boards-using-earle-philhower-arduino-pico-core)
    * [8.1. To use BOARD_NAME](#81-to-use-board_name)
    * [8.2. To avoid compile error relating to microsecondsToClockCycles](#82-to-avoid-compile-error-relating-to-microsecondstoclockcycles)
  * [9. For Portenta_H7 boards using Arduino IDE in Linux](#9-for-portenta_h7-boards-using-arduino-ide-in-linux)
  * [10. For RTL8720DN boards using AmebaD core](#10-for-rtl8720dn-boards-using-amebad-core)
  * [11. For SAMD21 and SAMD51 boards using ArduinoCore-fab-sam core](#11-For-SAMD21-and-SAMD51-boards-using-ArduinoCore-fab-sam-core)
  * [12. For Seeeduino RP2040 boards](#12-For-Seeeduino-RP2040-boards)
  * [13. For Seeeduino nRF52840 boards](#13-For-Seeeduino-nRF52840-boards)
* [Libraries' Patches](#libraries-patches)
  * [1. For application requiring 2K+ HTML page](#1-for-application-requiring-2k-html-page)
  * [2. How to use EthernetLarge feature](#2-How-to-use-EthernetLarge-feature)
  * [3. How to select another CS/SS pin to use](#3-how-to-select-another-csss-pin-to-use)
  * [4. How to use W5x00 with ESP8266](#4-how-to-use-w5x00-with-esp8266)
* [Ethernet_Generic Library API](#Ethernet_Generic-Library-API) 
* [Configuration Notes](#configuration-notes)
  * [1. How to select which SPI to use](#1-How-to-select-which-SPI-to-use)
  	* [1.1. For boards other than STM32](#11-For-boards-other-than-STM32)
    * [1.2. For STM32 boards](#12-For-STM32-boards)
  * [2. How to use EthernetLarge feature](#2-How-to-use-EthernetLarge-feature)
  * [3. How to select another CS/SS pin to use](#3-How-to-select-another-CSSS-pin-to-use)
  * [4. How to use W5x00 with ESP8266](#4-How-to-use-W5x00-with-ESP8266)
  * [5. Important Note for AVRDx using Arduino IDE](#5-Important-Note-for-AVRDx-using-Arduino-IDE) **New**
* [Examples](#examples)
  * [ 1. multiFileProject](examples/multiFileProject)
  * [ 2. TelnetClient](examples/TelnetClient)
  * [ 3. UdpNTPClient](examples/UdpNTPClient)
  * [ 4. UdpSendReceive](examples/UdpSendReceive)
  * [ 5. WebClient](examples/WebClient)
  * [ 6. WebClient_ESP](examples/WebClient_ESP)
  * [ 7. WebClient_ESP_SPI2](examples/WebClient_ESP_SPI2)
  * [ 8. WebClientRepeating](examples/WebClientRepeating)
  * [ 9. WebClientRepeating_ESP](examples/WebClientRepeating_ESP)
  * [10. WebClientRepeating_ESP_SPI2](examples/WebClientRepeating_ESP_SPI2)
  * [11. WebServer](examples/WebServer)
  * [12. WebClientRepeating_RP2040_SPI1](examples/WebClientRepeating_RP2040_SPI1) **New**
  * [13. SetDHCPHostName](examples/SetDHCPHostName) **New**
  * [14. EthernetWebServer_BigData](examples/EthernetWebServer_BigData) **New**
* [Example WebClientRepeating](#example-WebClientRepeating)
  * [1. File WebClientRepeating.ino](#1-file-WebClientRepeatingino)
  * [2. File defines.h](#2-file-definesh) 
* [Debug Terminal Output Samples](#debug-terminal-output-samples)
  * [ 1. WebClientRepeating_ESP on ESP32_DEV](#1-WebClientRepeating_ESP-on-ESP32_DEV)
  * [ 2. WebClientRepeating_ESP_SPI2 on ESP32_DEV](#2-WebClientRepeating_ESP_SPI2-on-ESP32_DEV)
  * [ 3. WebClientRepeating on AVR Mega](#3-WebClientRepeating-on-AVR-Mega)
  * [ 4. WebClientRepeating on NRF52840_FEATHER](#4-WebClientRepeating-on-NRF52840_FEATHER)
  * [ 5. WebClientRepeating on SAM DUE](#5-WebClientRepeating-on-SAM-DUE)
  * [ 6. WebClientRepeating on SAMD](#6-WebClientRepeating-on-SAMD)
  * [ 7. WebClientRepeating on NUCLEO_F767ZI](#7-WebClientRepeating-on-NUCLEO_F767ZI)
  * [ 8. UdpNTPClient on AVR Mega](#8-UdpNTPClient-on-AVR-Mega)
  * [ 9. WebClient on MBED RASPBERRY_PI_PICO with Large Buffer](#9-WebClient-on-MBED-RASPBERRY_PI_PICO-with-Large-Buffer)
  * [10. WebClient on RASPBERRY_PI_PICO with Large Buffer](#10-WebClient-on-RASPBERRY_PI_PICO-with-Large-Buffer)
  * [11. WebClientRepeating_RP2040_SPI1 on RASPBERRY_PI_PICO with Large Buffer](#11-WebClientRepeating_RP2040_SPI1-on-RASPBERRY_PI_PICO-with-Large-Buffer)
  * [12. WebClient on MBED RASPBERRY_PI_PICO with W5100S using Ethernet_Generic Library with Large Buffer](#12-WebClient-on-MBED-RASPBERRY_PI_PICO-with-W5100S-using-Ethernet_Generic-Library-with-Large-Buffer)
  * [13. WebClientRepeating on NUCLEO_L552ZE_Q with W5x00 using Ethernet_Generic Library with Large Buffer](#13-WebClientRepeating-on-NUCLEO_L552ZE_Q-with-W5x00-using-Ethernet_Generic-Library-with-Large-Buffer)
  * [14. WebClientRepeating on NUCLEO_F767ZI with W5x00 using Ethernet_Generic Library with Large Buffer](#14-WebClientRepeating-on-NUCLEO_F767ZI-with-W5x00-using-Ethernet_Generic-Library-with-Large-Buffer)
  * [15. WebClientRepeating_RP2040_SPI1 on MBED RASPBERRY_PI_PICO with W5x00 using Ethernet_Generic Library with Large Buffer](#15-WebClientRepeating_RP2040_SPI1-on-MBED-RASPBERRY_PI_PICO-with-W5x00-using-Ethernet_Generic-Library-with-Large-Buffer)
  * [16. SetDHCPHostName on WIZNET_5100S_EVB_PICO with W5x00 using Ethernet_Generic Library with Large Buffer](#16-SetDHCPHostName-on-WIZNET_5100S_EVB_PICO-with-W5x00-using-Ethernet_Generic-Library-with-Large-Buffer)
  * [17. WebClientRepeating on WIZNET_5500_EVB_PICO with W5x00 using Ethernet_Generic Library with Large Buffer](#17-WebClientRepeating-on-WIZNET_5500_EVB_PICO-with-W5x00-using-Ethernet_Generic-Library-with-Large-Buffer)
  * [18. WebClientRepeating on SAMD_FEATHER_M0_EXPRESS with W5x00 using Ethernet_Generic Library with Large Buffer](#18-WebClientRepeating-on-SAMD_FEATHER_M0_EXPRESS-with-W5x00-using-Ethernet_Generic-Library-with-Large-Buffer)
  * [19. UdpNTPClient on SAMD_ZERO with W5x00 using Ethernet_Generic Library with Large Buffer](#19-UdpNTPClient-on-SAMD_ZERO-with-W5x00-using-Ethernet_Generic-Library-with-Large-Buffer)
  * [20. WebClientRepeating on Curiosity AVR_AVR128DA48 with W5x00 using Ethernet_Generic Library with Large Buffer](#20-WebClientRepeating-on-Curiosity-AVR_AVR128DA48-with-W5x00-using-Ethernet_Generic-Library-with-Large-Buffer)
  * [21. EthernetWebServer_BigData on NRF52840_FEATHER](#21-EthernetWebServer_BigData-on-NRF52840_FEATHER)
  * [22. EthernetWebServer_BigData on RASPBERRY_PI_PICO](#22-EthernetWebServer_BigData-on-RASPBERRY_PI_PICO)
  * [23. WebClient_ESP on ESP32_DEV with W6100 using Ethernet_Generic Library on SPI](#23-WebClient_ESP-on-ESP32_DEV-with-W6100-using-Ethernet_Generic-Library-on-SPI)
* [Debug](#debug)
* [Troubleshooting](#troubleshooting)
* [Issues](#issues)
* [TO DO](#to-do)
* [DONE](#done)
* [Contributions and Thanks](#contributions-and-thanks)
* [Contributing](#contributing)
* [Copyright](#copyright)


---
---

### Important Note from v2.5.2

The new [Ethernet_Generic v2.8.1+](https://github.com/khoih-prog/Ethernet_Generic/releases/tag/v2.7.0) fixes severe limitation to permit sending much larger data than total than **2/4/8/16K** buffer of **W5x00 Ethernet**.

---

To be safe for systems using old `W5100` shields, the **examples** are using `optional` conservative values for SPI clock speed of 14MHz and `SPI_MODE0`

For `SAMD21 M0`, such as `SAMD Zero`, SPI clock speed of 8MHz and `SPI_MODE0` are used.


```cpp
// Default to use W5100. Must change to false for W5500, W5100S, for faster SPI clock
//#define USE_W5100                           true
```

To use with shields different from `W5100`, such as `W5200, W5500, W5100S, W6100`, change to 

```cpp
// Default to use W5100. Must change to false for W5500, W5100S, for faster SPI clock
#define USE_W5100                           false
```

For Arduino SAMD21 Zero, in order to print to Terminal, use `SERIAL_PORT_USBVIRTUAL` == `SerialUSB`

```cpp
// Use this for ARDUINO_SAMD_ZERO, etc. if can't print to terminal with Serial.print
#if defined(SERIAL_PORT_USBVIRTUAL)
  #define Serial          SERIAL_PORT_USBVIRTUAL
  #warning Using SAMD Zero SerialUSB
#endif
```

---

### Why do we need this [Ethernet_Generic library](https://github.com/khoih-prog/Ethernet_Generic)

#### Features

This [**Ethernet_Generic library**](https://github.com/khoih-prog/Ethernet_Generic) is a port and modification to merge, as many as possible, the features of the following libraries
1. [**Ethernet**](https://github.com/arduino-libraries/Ethernet)
2. [**EthernetLarge**](https://github.com/OPEnSLab-OSU/EthernetLarge)
3. [**Ethernet2**](https://github.com/adafruit/Ethernet2)
4. [**Ethernet3**](https://github.com/sstaub/Ethernet3)

for **AVR, Portenta_H7, Teensy, SAM DUE, Arduino SAMD21, Adafruit SAMD21/SAMD51, Adafruit nRF52, ESP32/ESP8266, STM32, RP2040-based, etc.** boards using Ethernet W5x00 shields. 

The `h-only` library now supports `SPI1/SPI2`, most of `EthernetLarge`, `Ethernet2` and `Ethernet3` features by simply changing definitions in the sketch.

---

#### Currently supported Boards

This [**Ethernet_Generic** library](https://github.com/khoih-prog/Ethernet_Generic) currently supports these following boards:

 1. **nRF52 boards**, such as **AdaFruit Feather nRF52832, nRF52840 Express, BlueFruit Sense, Itsy-Bitsy nRF52840 Express, Metro nRF52840 Express, NINA_B302_ublox, NINA_B112_ublox, etc.**
 2. **SAM DUE**
 3. **SAMD21**
  - Arduino SAMD21: ZERO, MKRs, NANO_33_IOT, etc.
  - Adafruit SAMD21 (M0): ItsyBitsy M0, Feather M0, Feather M0 Express, Metro M0 Express, Circuit Playground Express, Trinket M0, PIRkey, Hallowing M0, Crickit M0, etc.
  - Seeeduino:  LoRaWAN, Zero, Femto M0, XIAO M0, Wio GPS Board, etc.
  
 4. **SAMD51**
  - Adafruit SAMD51 (M4): Metro M4, Grand Central M4, ItsyBitsy M4, Feather M4 Express, Trellis M4, Metro M4 AirLift Lite, MONSTER M4SK Express, Hallowing M4, etc.
  - Seeeduino: Wio Terminal, Grove UI Wireless
  
 5. **Teensy (4.1, 4.0, 3.6, 3.5, 3,2, 3.1, 3.0, LC)**
 6. **AVR Mega1280, 2560, ADK.**
 7. ESP32, using **SPI or SPI2**
 8. ESP8266
 
 9. RP2040-based boards, such as **RASPBERRY_PI_PICO, ADAFRUIT_FEATHER_RP2040 and GENERIC_RP2040**, using [**Arduino-mbed RP2040** core](https://github.com/arduino/ArduinoCore-mbed) with SPI, SPI1, or custom SPI. **New**
10. RP2040-based boards, such as **RASPBERRY_PI_PICO, ADAFRUIT_FEATHER_RP2040 and GENERIC_RP2040**, using [**Earle Philhower's arduino-pico** core](https://github.com/earlephilhower/arduino-pico) with SPI, SPI1 or custom SPI. **New**
11. **Portenta_H7** with SPI0, SPI1 or custom SPI. **New**
 
12. **Arduino UNO WiFi Rev2, AVR_NANO_EVERY, etc.**
  
13. **STM32F/L/H/G/WB/MP1 boards (with 32+K Flash) running W5x00 shields)** with SPI, SPI2 or custom SPI. **New**

- Nucleo-144
- Nucleo-64
- Discovery
- Generic STM32F0, STM32F1, STM32F2, STM32F3, STM32F4, STM32F7 (with 64+K Flash): x8 and up
- STM32L0, STM32L1, STM32L4
- STM32G0, STM32G4
- STM32H7
- STM32WB
- STM32MP1
- LoRa boards
- 3-D printer boards
- Generic Flight Controllers
- Midatronics boards

12. **Arduino AVR_Dx boards using DxCore**

- **AVRDA-based boards (AVR128DA, AVR64DA, AVR32DA, etc.)**

<p align="center">
    <img src="https://github.com/khoih-prog/Ethernet_Generic/raw/main/pics/Curiosity_AVR128DA48.png">
</p>


- **AVRDB-based boards (AVR128DB, AVR64DB, AVR32DB, etc.)**

<p align="center">
    <img src="https://github.com/khoih-prog/Ethernet_Generic/raw/main/pics/Curiosity_AVR128DB48.png">
</p>


13. **Seeeduino nRF52840-based boards such as XIAO_NRF52840 and XIAO_NRF52840_SENSE**, etc. using Seeed `mbed` or `nRF52` core


#### Currently supported Ethernet shields/modules

1. W5x00 shield /module such as W5100, W5200, W5500 and W5100S
2. W5100S shield /module, such as [**WIZnet Ethernet HAT**](https://docs.wiznet.io/Product/Open-Source-Hardware/wiznet_ethernet_hat) and [**W5100S-EVB-Pico**](https://docs.wiznet.io/Product/iEthernet/W5100S/w5100s-evb-pico)
3. W6100 shield /module, such as [**WIZnet WIZ610MJ**](https://eshop.wiznet.io/shop/module/network-module/wiz610mj/) 


<p align="center">
    <img src="https://github.com/khoih-prog/Ethernet_Generic/raw/main/pics/W6100.png">
</p>


---
---


## Prerequisites

 1. [`Arduino IDE 1.8.19+` for Arduino](https://github.com/arduino/Arduino). [![GitHub release](https://img.shields.io/github/release/arduino/Arduino.svg)](https://github.com/arduino/Arduino/releases/latest)
 2. [`Arduino AVR core 1.8.6+`](https://github.com/arduino/ArduinoCore-avr) for Arduino (Use Arduino Board Manager) AVR boards. [![GitHub release](https://img.shields.io/github/release/arduino/ArduinoCore-avr.svg)](https://github.com/arduino/ArduinoCore-avr/releases/latest)
 3. [`Teensy core v1.57+`](https://github.com/PaulStoffregen/cores) for Teensy (4.1, 4.0, 3.6, 3.5, 3,2, 3.1, 3.0) boards. [![GitHub release](https://img.shields.io/github/release/PaulStoffregen/cores.svg)](https://github.com/PaulStoffregen/cores/releases/latest)
 4. [`Arduino SAM DUE core v1.6.12+`](https://github.com/arduino/ArduinoCore-sam) for SAM DUE ARM Cortex-M3 boards.
 5. [`Arduino SAMD core 1.8.13+`](https://github.com/arduino/ArduinoCore-samd) for SAMD ARM Cortex-M0+ boards. [![GitHub release](https://img.shields.io/github/release/arduino/ArduinoCore-samd.svg)](https://github.com/arduino/ArduinoCore-samd/releases/latest)
 6. [`Adafruit SAMD core 1.7.11+`](https://github.com/adafruit/ArduinoCore-samd) for SAMD ARM Cortex-M0+ and M4 boards (Nano 33 IoT, etc.). [![GitHub release](https://img.shields.io/github/release/adafruit/ArduinoCore-samd.svg)](https://github.com/adafruit/ArduinoCore-samd/releases/latest)
 7. [`Seeeduino SAMD core 1.8.3+`](https://github.com/Seeed-Studio/ArduinoCore-samd) for SAMD21/SAMD51 boards (XIAO M0, Wio Terminal, etc.). [![Latest release](https://img.shields.io/github/release/Seeed-Studio/ArduinoCore-samd.svg)](https://github.com/Seeed-Studio/ArduinoCore-samd/releases/latest/)
 8. [`Adafruit nRF52 v1.3.0+`](https://github.com/adafruit/Adafruit_nRF52_Arduino) for nRF52 boards such as Adafruit NRF52840_FEATHER, NRF52832_FEATHER, NRF52840_FEATHER_SENSE, NRF52840_ITSYBITSY, NRF52840_CIRCUITPLAY, NRF52840_CLUE, NRF52840_METRO, NRF52840_PCA10056, PARTICLE_XENON, **NINA_B302_ublox**, etc. [![GitHub release](https://img.shields.io/github/release/adafruit/Adafruit_nRF52_Arduino.svg)](https://github.com/adafruit/Adafruit_nRF52_Arduino/releases/latest)
 9. [`ESP32 Core 2.0.6+`](https://github.com/espressif/arduino-esp32) for ESP32-based boards. [![Latest release](https://img.shields.io/github/release/espressif/arduino-esp32.svg)](https://github.com/espressif/arduino-esp32/releases/latest/)
10. [`ESP8266 Core 3.1.0+`](https://github.com/esp8266/Arduino) for ESP8266-based boards. [![Latest release](https://img.shields.io/github/release/esp8266/Arduino.svg)](https://github.com/esp8266/Arduino/releases/latest/). To use ESP8266 core 2.7.1+ for LittleFS.

changelog.md
## Ethernet_Generic

[![arduino-library-badge](https://www.ardu-badge.com/badge/Ethernet_Generic.svg?)](https://www.ardu-badge.com/Ethernet_Generic)
[![GitHub release](https://img.shields.io/github/release/khoih-prog/Ethernet_Generic.svg)](https://github.com/khoih-prog/Ethernet_Generic/releases)
[![contributions welcome](https://img.shields.io/badge/contributions-welcome-brightgreen.svg?style=flat)](#Contributing)
[![GitHub issues](https://img.shields.io/github/issues/khoih-prog/Ethernet_Generic.svg)](http://github.com/khoih-prog/Ethernet_Generic/issues)

<a href="https://www.buymeacoffee.com/khoihprog6" title="Donate to my libraries using BuyMeACoffee"><img src="https://cdn.buymeacoffee.com/buttons/v2/default-yellow.png" alt="Donate to my libraries using BuyMeACoffee" style="height: 50px !important;width: 181px !important;" ></a>
<a href="https://www.buymeacoffee.com/khoihprog6" title="Donate to my libraries using BuyMeACoffee"><img src="https://img.shields.io/badge/buy%20me%20a%20coffee-donate-orange.svg?logo=buy-me-a-coffee&logoColor=FFDD00" style="height: 20px !important;width: 200px !important;" ></a>


---
---

## Table of Contents


* [Changelog](#changelog)
	* [Releases v2.8.1](#releases-v281)
	* [Releases v2.8.0](#releases-v280)
	* [Releases v2.7.1](#releases-v271)
	* [Releases v2.7.0](#releases-v270)
	* [Releases v2.6.2](#releases-v262)
	* [Releases v2.6.1](#releases-v261)
	* [Releases v2.6.0](#releases-v260)
	* [Releases v2.5.2](#releases-v252)
	* [Releases v2.5.1](#releases-v251)
	* [Releases v2.5.0](#releases-v250)
	* [Releases v2.4.1](#releases-v241)
	* [Releases v2.4.0](#releases-v240)
	* [Releases v2.3.1](#releases-v231)
	* [Releases v2.3.0](#releases-v230)
	* [Releases v2.2.0](#releases-v220)
	* [Releases v2.1.0](#releases-v210)
	* [Releases v2.0.1](#releases-v201)
	* [Releases v2.0.0](#releases-v200)

---
---

## Changelog

### Releases v2.8.1

1. Fix minor `speed-link-duplex` bug for `WIZNet W6100`

### Releases v2.8.0

1. Add preliminary support to `WIZNet W6100` using IPv4. IPv6 will be supported in the future

### Releases v2.7.1

1. Auto-detect `W5x00` and settings (`ETHERNET_LARGE_BUFFERS`, `MAX_SOCK_NUM`) to set `MAX_SIZE` to send. Check ['ETHERNET_SERVER_SEND_MAX_SIZE' was not declared in this scope #16](https://github.com/khoih-prog/Ethernet_Generic/issues/16)


### Releases v2.7.0

1. Fix severe limitation to permit sending larger data than 2/4/8/16K buffer
2. Add examples [EthernetWebServer_BigData](https://github.com/khoih-prog/Ethernet_Generic/tree/main/examples/EthernetWebServer_BigData) to demo how to send larger data than 2/4/8/16K buffer
3. Optimize code
4. Clean up
5. Update `Packages' Patches`

### Releases v2.6.2

1. Add support to Seeeduino nRF52840-based boards such as **Seeed XIAO_NRF52840 and XIAO_NRF52840_SENSE**, etc. using Seeed `mbed` or `nRF52` core
2. Add astyle using `allman` style. Restyle the library
3. Update `Packages' Patches` to add Seeeduino `nRF52` core

### Releases v2.6.1

1. Fix bug for `Wiznet W5200 / wiz820io`. Check [ESP32 - Wiznet W5200 problem #13](https://github.com/khoih-prog/Ethernet_Generic/discussions/13)
2. Update `Packages' Patches`

### Releases v2.6.0

1. Add support to AVR Dx (AVR128Dx, AVR64Dx, AVR32Dx, etc.) using [DxCore](https://github.com/SpenceKonde/DxCore)


### Releases v2.5.2

1. Slow SPI clock only when necessary (such as for W5100, SAMD21 Zero, etc.)
2. Use correct Debug Terminal `Serial` for so-called **SAMD21 Zero** boards from Arduino as well as Adafruit


### Releases v2.5.1

1. Slow SPI clock to `8MHz` and select `SPI_MODE0` for old W5100 shield using SAMD Zero. Check [AdvancedWebServer example give me wrong ip HTTP EthernetWebServer is @ IP : 0.0.0.0 #62](https://github.com/khoih-prog/EthernetWebServer/issues/62)
2. Default to `W5100` to be safe. To change if using `W5200, W5500, W5100s` or testing faster SPI clock speeds


### Releases v2.5.0

1. Using `raw_address()` as default only for arduino-pico core for compatibility with remaining cores
2. Sync with [Ethernet library v2.0.1](https://github.com/arduino-libraries/Ethernet/releases/tag/2.0.1)
3. Add [Ethernet_Generic Library API](https://github.com/khoih-prog/Ethernet_Generic/blob/main/docs/api.md)


### Releases v2.4.1

1. Auto-select SPI(s) `SS/CS` pins according to board package if available
2. Update `Packages' Patches`

### Releases v2.4.0

1. Using raw_address() as default instead of private IPAddress data
2. Add missing `LibraryPatches` for ESP32
3. Update `Packages' Patches`

### Releases v2.3.1

1. Add setHostname() and related DHCP functions. Check [Only 15 UDP packets / second with W5100S-EVB-Pico possible? #6](https://github.com/khoih-prog/Ethernet_Generic/discussions/6#discussioncomment-2793483)
2. Add example [SetDHCPHostName](examples/SetDHCPHostName) to demo the new feature
3. Update `Packages' Patches`

### Releases v2.3.0

1. Add support to custom SPI for RP2040, Nano_RP2040_Connect, Portenta_H7, etc. using [Arduino-mbed core](https://github.com/arduino/ArduinoCore-mbed)


### Releases v2.2.0

1. Add support to STM32 boards using [Arduino_Core_STM32](https://github.com/stm32duino/Arduino_Core_STM32)
2. Add support to any board with core supporting custom hardware / software `SPI`, for example STM32.


### Releases v2.1.0

1. Add support to WIZNet W5100S. Check [Wiznet W5100S #56](https://github.com/khoih-prog/EthernetWebServer/discussions/56)


### Releases v2.0.1

1. Add support to `SPI1` for `RP2040` using [`Earle Philhower's arduino-pico` core](https://github.com/earlephilhower/arduino-pico)
2. Add example [WebClientRepeating_RP2040_SPI1](examples/WebClientRepeating_RP2040_SPI1) to demo the new feature
3. Update `Packages' Patches`

### Releases v2.0.0

1. Initial porting and coding to merge, as many as possible, the features of the following libraries :

	- [Ethernet](https://github.com/arduino-libraries/Ethernet)
	- [EthernetLarge](https://github.com/OPEnSLab-OSU/EthernetLarge)

src/Dhcp.h
/****************************************************************************************************************************
  Dhcp.h - Driver for W5x00

  Ethernet_Generic is a library for the W5x00 Ethernet shields trying to merge the good features of
  previous Ethernet libraries

  Based on and modified from

  1) Ethernet Library         https://github.com/arduino-libraries/Ethernet
  2) EthernetLarge Library    https://github.com/OPEnSLab-OSU/EthernetLarge
  3) Ethernet2 Library        https://github.com/adafruit/Ethernet2
  4) Ethernet3 Library        https://github.com/sstaub/Ethernet3

  Built by Khoi Hoang https://github.com/khoih-prog/EthernetWebServer

  Version: 2.8.1

  Version Modified By   Date      Comments
  ------- -----------  ---------- -----------
  2.0.0   K Hoang      31/03/2022 Initial porting and coding to support SPI2, debug, h-only library
  ...
  2.6.0   K Hoang      11/09/2022 Add support to AVR Dx (AVR128Dx, AVR64Dx, AVR32Dx, etc.) using DxCore
  2.6.1   K Hoang      23/09/2022 Fix bug for W5200
  2.6.2   K Hoang      26/10/2022 Add support to Seeed XIAO_NRF52840 and XIAO_NRF52840_SENSE using `mbed` or `nRF52` core
  2.7.0   K Hoang      14/11/2022 Fix severe limitation to permit sending larger data than 2/4/8/16K buffer
  2.7.1   K Hoang      15/11/2022 Auto-detect W5x00 and settings to set MAX_SIZE to send
  2.8.0   K Hoang      27/12/2022 Add support to W6100 using IPv4
  2.8.1   K Hoang      06/01/2023 Fix W6100 minor bug
 *****************************************************************************************************************************/

// DHCP Library v0.3 - April 25, 2009
// Author: Jordan Terrell - blog.jordanterrell.com

#ifndef Dhcp_h
#define Dhcp_h

/* DHCP state machine. */
#define STATE_DHCP_START        0
#define STATE_DHCP_DISCOVER     1
#define STATE_DHCP_REQUEST      2
#define STATE_DHCP_LEASED       3
#define STATE_DHCP_REREQUEST    4
#define STATE_DHCP_RELEASE      5

#define DHCP_FLAGSBROADCAST     0x8000

/* UDP port numbers for DHCP */
#define DHCP_SERVER_PORT        67  /* from server to client */
#define DHCP_CLIENT_PORT        68  /* from client to server */

/* DHCP message OP code */
#define DHCP_BOOTREQUEST        1
#define DHCP_BOOTREPLY          2

/* DHCP message type */
#define DHCP_DISCOVER           1
#define DHCP_OFFER              2
#define DHCP_REQUEST            3
#define DHCP_DECLINE            4
#define DHCP_ACK                5
#define DHCP_NAK                6
#define DHCP_RELEASE            7
#define DHCP_INFORM             8

#define DHCP_HTYPE10MB          1
#define DHCP_HTYPE100MB         2

#define DHCP_HLENETHERNET       6

#define DHCP_HOPS               0
#define DHCP_SECS               0

#define MAGIC_COOKIE            0x63825363
#define MAX_DHCP_OPT            16

#define HOST_NAME               "WIZnet"
#define DEFAULT_LEASE           (900) //default lease time in seconds

#define DHCP_CHECK_NONE         (0)
#define DHCP_CHECK_RENEW_FAIL   (1)
#define DHCP_CHECK_RENEW_OK     (2)
#define DHCP_CHECK_REBIND_FAIL  (3)
#define DHCP_CHECK_REBIND_OK    (4)

enum
{
  padOption                 = 0,
  subnetMask                = 1,
  timerOffset               = 2,
  routersOnSubnet           = 3,

  /* timeServer             = 4,
    nameServer              = 5,*/

  dns                       = 6,

  /*logServer               = 7,
    cookieServer            = 8,
    lprServer               = 9,
    impressServer           = 10,
    resourceLocationServer  = 11,*/

  hostName                  = 12,

  /*bootFileSize            = 13,
    meritDumpFile           = 14,*/

  domainName                = 15,

  /*swapServer              = 16,
    rootPath                = 17,
    extentionsPath          = 18,
    IPforwarding            = 19,
    nonLocalSourceRouting   = 20,
    policyFilter            = 21,
    maxDgramReasmSize       = 22,
    defaultIPTTL            = 23,
    pathMTUagingTimeout     = 24,
    pathMTUplateauTable     = 25,
    ifMTU                   = 26,
    allSubnetsLocal         = 27,
    broadcastAddr           = 28,
    performMaskDiscovery    = 29,
    maskSupplier            = 30,
    performRouterDiscovery  = 31,
    routerSolicitationAddr  = 32,
    staticRoute             = 33,
    trailerEncapsulation    = 34,
    arpCacheTimeout         = 35,
    ethernetEncapsulation   = 36,
    tcpDefaultTTL           = 37,
    tcpKeepaliveInterval    = 38,
    tcpKeepaliveGarbage     = 39,
    nisDomainName           = 40,
    nisServers              = 41,
    ntpServers              = 42,
    vendorSpecificInfo      = 43,
    netBIOSnameServer       = 44,
    netBIOSdgramDistServer  = 45,
    netBIOSnodeType         = 46,
    netBIOSscope            = 47,
    xFontServer             = 48,
    xDisplayManager         = 49,*/

  dhcpRequestedIPaddr       = 50,
  dhcpIPaddrLeaseTime       = 51,

  /*dhcpOptionOverload      = 52,*/

  dhcpMessageType           = 53,
  dhcpServerIdentifier      = 54,
  dhcpParamRequest          = 55,

  /*dhcpMsg                 = 56,
    dhcpMaxMsgSize          = 57,*/

  dhcpT1value               = 58,
  dhcpT2value               = 59,

  /*dhcpClassIdentifier     = 60,*/

  dhcpClientIdentifier      = 61,
  endOption                 = 255
};

typedef struct _RIP_MSG_FIXED
{
  uint8_t  op;
  uint8_t  htype;
  uint8_t  hlen;
  uint8_t  hops;
  uint32_t xid;
  uint16_t secs;
  uint16_t flags;
  uint8_t  ciaddr[4];
  uint8_t  yiaddr[4];
  uint8_t  siaddr[4];
  uint8_t  giaddr[4];
  uint8_t  chaddr[6];
} RIP_MSG_FIXED;

#endif
